db_test.go

```diff
@@ -935,6 +935,55 @@ func TestDB_Retention(t *testing.T) {
 	testutil.Equals(t, int64(100), db.blocks[0].meta.MaxTime) // To verify its the right block.
 }
 
+func TestSizeBasedRetention(t *testing.T) {
+	db, close := openTestDB(t, &Options{
+		BlockRanges: []int64{100},
+	})
+	defer close()
+
+	// Add some empty blocks.
+	blocks := []*BlockMeta{
+		{ULID: ulid.MustNew(100, nil), MinTime: 100, MaxTime: 200},
+		{ULID: ulid.MustNew(200, nil), MinTime: 200, MaxTime: 300},
+		{ULID: ulid.MustNew(300, nil), MinTime: 300, MaxTime: 400},
+		{ULID: ulid.MustNew(400, nil), MinTime: 400, MaxTime: 500},
+		{ULID: ulid.MustNew(500, nil), MinTime: 500, MaxTime: 600},
+	}
+	for _, m := range blocks {
+		createEmptyBlock(t, filepath.Join(db.Dir(), m.ULID.String()), m)
+	}
+
+	// Figure out how large an empty block is.
+	emptyBlockDir := filepath.Join(db.Dir(), blocks[0].ULID.String())
+	emptyBlock, err := OpenBlock(emptyBlockDir, db.chunkPool)
+	testutil.Ok(t, err)
+	db.opts.MaxBytes = 4 * emptyBlock.Size()
+
+	// Reload and ensure that actual size is less than limit.
+	testutil.Ok(t, db.reload())
+
+	var size int64
+	for _, b := range db.blocks {
+		size += b.Size()
+	}
+	testutil.Assert(t, size <= db.opts.MaxBytes, "actual size (%v) is expected to be less than or equal to limit (%v)", size, db.opts.MaxBytes)
+	testutil.Equals(t, 4, len(db.blocks))
+	testutil.Equals(t, int64(300), db.blocks[0].meta.MaxTime) // Ensure oldest block was deleted.
+
+	// Reduce the size limit and test again.
+	db.opts.MaxBytes = 1 * emptyBlock.Size()
```

Comment:
why not just  : `db.opts.MaxBytes = emptyBlock.Size()` or `db.opts.MaxBytes -= emptyBlock.Size()`  :stuck_out_tongue_winking_eye:

Reply:
Yeah, I'm not exactly sure why I did that haha